### PR TITLE
RHOAIENG-183: Use ODH OpenVino image, and other fixes

### PR DESCRIPTION
**Pull request by grdryn**
JIRA: https://issues.redhat.com/browse/RHOAIENG-183

## Description

- Use ODH OpenVino image in base Containerfile
The source for that image is available at https://github.com/opendatahub-io/openvino_model_server
I also squashed a coupld of RUN commands into one in that Containerfile.
- Use a consistent default for MODEL_DIR in Containerfiles
I don't know whether it's the best value, but the previous on in the OpenVino Containerfile seemed too specific to the tensorflow example model.
- This makes it easier to use the same example curl command to test that the image built correctly, whether from our Pipeline or using the example `podman run` in this README.
  - Use ODH OpenVino base image
  - Use same REST port as pipeline Containerfiles use (8080)
  - Use onnx-mnist as the name rather than onnx_mnist
  - Use Podman in examples 🦭
- Add OpenVino PipelineRun example for aiedge-e2e
This also renames the pre-existing one to make it clear that it's for the bike-rentals example, and updates the README to mention both.

## How Has This Been Tested?

- The included PipelineRun file for the tensorflow-housing example was created to test this change. Additionally, I ran the secondary example models locally with Podman.

## Merge criteria:

- [x] The commits are squashed in a cohesive manner and have meaningful messages.
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

File: pipelines/models/tensorflow-facedetection/README.md
Comment (grdryn) [inline pipelines/models/tensorflow-facedetection/README.md:21]: I actually couldn't get this particular test request to work...I had issues with Python dependencies.

File: pipelines/containerfiles/Containerfile.openvino.mlserver.mlflow
Comment (grdryn) [reply pipelines/containerfiles/Containerfile.openvino.mlserver.mlflow:21]: @davidesalerno in my understanding, every RUN instruction in a Dockerfile results in an additional layer in the resulting image. Sometimes there are good reasons to split RUN instructions into separate layers, but if there isn't, I think it's a better practice to keep them in a single layer. This can avoid size bloat in cases where, for example, two commands get run and the second removes some temporary files that the first one creates. While that example doesn't apply to this particular case right now, I think it's a reasonable idea to group these commands into a single RUN, so that it might be more obvious to people making changes in the future, that they can just put more commands into it too, if it makes sense.

Counter question, to make sure I'm not missing anything: are there reasons that you can think of that we _shouldn't_ group these commands into a single RUN like this?

Also, with reference to the "rearrange" part of your question: they're not rearranged into a different order, they will all be executed in the same order that they were in previously.

Comment (MarianMacik) [reply pipelines/containerfiles/Containerfile.openvino.mlserver.mlflow:21]: Nice, didn't even know this is possible.

File: pipelines/models/onnx-mnist/README.md
Comment (MarianMacik) [inline pipelines/models/onnx-mnist/README.md:23]: This has been moved.
```suggestion
- https://github.com/onnx/models/tree/main/validated/vision/classification/mnist
```

File: pipelines/tekton/aiedge-e2e/aiedge-e2e.pipeline.yaml
Comment (MarianMacik) [inline pipelines/tekton/aiedge-e2e/aiedge-e2e.pipeline.yaml:35]: We need to revisit this in a separate PR/Jira as effectively we now have 2 relative paths :) But I think an issue for that already exists, I will check.

File: pipelines/tekton/aiedge-e2e/aiedge-e2e.tensorflow-housing.pipelinerun.yaml
Comment (MarianMacik) [inline pipelines/tekton/aiedge-e2e/aiedge-e2e.tensorflow-housing.pipelinerun.yaml:24]:
```suggestion
    value: "s3"
```
Or overwrite this in the pipeline tests.
Currently we test S3 using both examples as this is more important than Git use case AFAIK. There is a plan to include also Git testing, of course, but just for now I would stick with the git.

File: test/shell-pipeline-tests/openvino-tensorflow-housing/pipelines-test-openvino-tensorflow-housing.sh
Comment (MarianMacik) [inline test/shell-pipeline-tests/openvino-tensorflow-housing/pipelines-test-openvino-tensorflow-housing.sh:38]:
```suggestion
cp "$AIEDGE_E2E_PIPELINE_DIR_PATH"/aiedge-e2e.tensorflow-housing.pipelinerun.yaml "$AIEDGE_E2E_PIPELINE_OVERRIDDEN_PATH"
sed -i "s|value: rhoai-edge-models|value: rhoai-edge-models-ci|" "$AIEDGE_E2E_PIPELINE_OVERRIDDEN_PATH"
sed -i "s|value: pipelines/models/|value: \"\"|" "$AIEDGE_E2E_PIPELINE_OVERRIDDEN_PATH"
```
+ possibly git to s3 or you can do it in the pipelinerun directly.

File: pipelines/tekton/aiedge-e2e/aiedge-e2e.tensorflow-housing.pipelinerun.yaml
Comment (grdryn) [reply pipelines/tekton/aiedge-e2e/aiedge-e2e.tensorflow-housing.pipelinerun.yaml:24]: @MarianMacik what's the value of testing S3 in both examples? Wouldn't the fact that it's used in one cover that pipeline task, and using git in the other one be additional coverage for the git clone task? My assumption was that they were of equal/similar importance.

File: pipelines/tekton/aiedge-e2e/aiedge-e2e.pipeline.yaml
Comment (grdryn) [reply pipelines/tekton/aiedge-e2e/aiedge-e2e.pipeline.yaml:35]: Yes, as mentioned in the commit message for this one, it's kind of a hack because of something that was hardcoded in the Containerfile. IMHO it's a symptom of having one uber-generic pipeline to try to cover all types of models and model servers.

File: test/shell-pipeline-tests/openvino-tensorflow-housing/pipelines-test-openvino-tensorflow-housing.sh
Comment (grdryn) [reply test/shell-pipeline-tests/openvino-tensorflow-housing/pipelines-test-openvino-tensorflow-housing.sh:38]: As mentioned in the other comment (which I think is related to this, right?), I think having one test use a model from S3 and the other one use an example from Git would be better than having both from S3.

Comment (MarianMacik) [reply test/shell-pipeline-tests/openvino-tensorflow-housing/pipelines-test-openvino-tensorflow-housing.sh:38]: @grdryn Yes, possibly. When I was implementing these tests I didn't see into it that much. The question is whether we need to test both example models on S3 and Git or not. Can it happen that one example model will run fine when fetched from S3 but not from Git or vice versa?

One thing which may theoretically occur is that we somehow move model/rename it in this Git repo and we forget to copy the changes to the S3 or vice versa as S3 contains a copy of these models. However, this wouldn't be an issue of our pipeline, of course. I am just thinking out loud whether we want to have the full coverage of all combinations or not.

If we are fine with having one for S3 and the other for Git then you can set relative path to `""` for the s3 seldonio-bike-rentals and remove that sed line from the test.

Comment (grdryn) [reply test/shell-pipeline-tests/openvino-tensorflow-housing/pipelines-test-openvino-tensorflow-housing.sh:38]: > The question is whether we need to test both example models on S3 and Git or not. Can it happen that one example model will run fine when fetched from S3 but not from Git or vice versa?

In my opinion, the example models aren't the crucial thing that we want to verify the correctness of, it's the Pipeline and its Tasks. I don't think we would get any additional value from testing each model with each download method. By running one instance of the pipeline (for any model) with the S3 method, and another instance of the pipeline (for any model) with the Git method, then we're getting good coverage of the different Pipeline variations.
It's also good that the two examples we're using in the tests use different model servers, which tests the different Containerfiles that we have for those.

> One thing which may theoretically occur is that we somehow move model/rename it in this Git repo and we forget to copy the changes to the S3 or vice versa as S3 contains a copy of these models. However, this wouldn't be an issue of our pipeline, of course. I am just thinking out loud whether we want to have the full coverage of all combinations or not.

At that point, I think it's almost like what we'd be testing is the structure of the `rhoai-edge-models-ci` S3 bucket (which isn't an output of this project), rather than the pipeline or the example, right? We don't want to "overfit the data" (to borrow a term from machine learning) – as long as users make a PipelineRun that is correct for their model and where it's stored, our Pipeline should succeed. It shouldn't need to be completely like the structure or naming of our examples, as long as they can fit the Pipeline to it by configuring the parameters that we've made available.

> If we are fine with having one for S3 and the other for Git then you can set relative path to "" for the s3 seldonio-bike-rentals and remove that sed line from the test.

Done now, thanks! :+1:

Comment (MarianMacik) [reply test/shell-pipeline-tests/openvino-tensorflow-housing/pipelines-test-openvino-tensorflow-housing.sh:38]: Yes, you are right, I think this way it's both sufficient and more efficient. If it was a very short test it would be doable but it doesn't make sense to test every possible combination with such long-running tests.

File: Makefile
Comment (MarianMacik) [inline Makefile:40]: There is a conflict after the previous PR was merged.

**Review by MarianMacik — COMMENTED** [review]
Looks good to me. I think after the conflict is solved, I can give an approval. Unless we want to wait for others, of course.

File: Makefile
Comment (grdryn) [reply Makefile]: Fixed, thanks :+1:

**Comment by MarianMacik** [comment]
@davidesalerno @grdryn are we fine with approving this? If so, I can go ahead.

**Review by MarianMacik — APPROVED** [review]
Thanks!

**Pull request merged**
